pkg/resource/parameter_group/delta.go

```diff
@@ -0,0 +1,85 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Code generated by ack-generate. DO NOT EDIT.
+
+package parameter_group
+
+import (
+	"bytes"
+	"github.com/aws-controllers-k8s/memorydb-controller/pkg/common"
+	"reflect"
+
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+)
+
+// Hack to avoid import errors during build...
+var (
+	_ = &bytes.Buffer{}
+	_ = &reflect.Method{}
+)
+
+// newResourceDelta returns a new `ackcompare.Delta` used to compare two
+// resources
+func newResourceDelta(
+	a *resource,
+	b *resource,
+) *ackcompare.Delta {
+	delta := ackcompare.NewDelta()
+	if (a == nil && b != nil) ||
+		(a != nil && b == nil) {
+		delta.Add("", a, b)
+		return delta
+	}
+
+	if ackcompare.HasNilDifference(a.ko.Spec.Description, b.ko.Spec.Description) {
+		delta.Add("Spec.Description", a.ko.Spec.Description, b.ko.Spec.Description)
+	} else if a.ko.Spec.Description != nil && b.ko.Spec.Description != nil {
+		if *a.ko.Spec.Description != *b.ko.Spec.Description {
+			delta.Add("Spec.Description", a.ko.Spec.Description, b.ko.Spec.Description)
+		}
+	}
+	if ackcompare.HasNilDifference(a.ko.Spec.Family, b.ko.Spec.Family) {
+		delta.Add("Spec.Family", a.ko.Spec.Family, b.ko.Spec.Family)
+	} else if a.ko.Spec.Family != nil && b.ko.Spec.Family != nil {
+		if *a.ko.Spec.Family != *b.ko.Spec.Family {
+			delta.Add("Spec.Family", a.ko.Spec.Family, b.ko.Spec.Family)
+		}
+	}
+	if ackcompare.HasNilDifference(a.ko.Spec.Name, b.ko.Spec.Name) {
+		delta.Add("Spec.Name", a.ko.Spec.Name, b.ko.Spec.Name)
+	} else if a.ko.Spec.Name != nil && b.ko.Spec.Name != nil {
+		if *a.ko.Spec.Name != *b.ko.Spec.Name {
+			delta.Add("Spec.Name", a.ko.Spec.Name, b.ko.Spec.Name)
+		}
+	}
+	if !reflect.DeepEqual(a.ko.Spec.ParameterNameValues, b.ko.Spec.ParameterNameValues) {
+		delta.Add("Spec.ParameterNameValues", a.ko.Spec.ParameterNameValues, b.ko.Spec.ParameterNameValues)
+	}
+	if !reflect.DeepEqual(a.ko.Spec.Tags, b.ko.Spec.Tags) {
+		delta.Add("Spec.Tags", a.ko.Spec.Tags, b.ko.Spec.Tags)
+	}
+
+	if delta.DifferentAt("Spec.Description") {
+		common.RemoveFromDelta(delta, "Spec.Description")
+	}
+
+	if delta.DifferentAt("Spec.Name") {
+		common.RemoveFromDelta(delta, "Spec.Name")
+	}
+
+	if delta.DifferentAt("Spec.Family") {
+		common.RemoveFromDelta(delta, "Spec.Family")
+	}
```

Comment:
I am not sure about this. This will allow the reconciliation to be successful even when the desired and latest objects will not be same.

IMO, any time user updates these three fields, an error should be returned pointing them to the fact that these fields are immutable.

Reply:
For now I am ignoring this will work on `Advisory` condition later